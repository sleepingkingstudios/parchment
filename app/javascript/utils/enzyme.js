import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
