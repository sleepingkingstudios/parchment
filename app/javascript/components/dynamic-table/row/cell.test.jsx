import React from 'react';
import { shallow } from 'enzyme';

import DynamicTableRowCell from './cell';

describe('<DynamicTableRowCell />', () => {
  const data = {
    id: '00000000-0000-0000-0000-000000000000',
    name: 'Self-sealing Stem Bolt',
    price: '10.00',
    quantity: 50,
  };
  const defaultProps = { data, label: true, prop: 'name' };

  describe('with default properties', () => {
    it('should render the row cell', () => {
      const rendered = shallow(<DynamicTableRowCell {...defaultProps} />);
      const { name } = data;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('col');
      expect(rendered).toHaveText(name);
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<DynamicTableRowCell {...defaultProps} />);

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with resourceName: value', () => {
    const resourceName = 'widgets';

    it('should render the row cell', () => {
      const rendered = shallow(
        <DynamicTableRowCell {...defaultProps} resourceName={resourceName} />,
      );
      const { name } = data;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('widgets-table-row-cell');
      expect(rendered).toHaveClassName('widgets-table-row-name-cell');
      expect(rendered).toHaveClassName('col');
      expect(rendered).toHaveText(name);
    });
  });

  describe('with value: a component', () => {
    const CellValue = (props) => {
      const { price, quantity } = props;

      return (
        <span>{ quantity } available at ${ price } each</span>
      );
    };

    it('should render the row cell', () => {
      const rendered = shallow(<DynamicTableRowCell {...defaultProps} prop="price" value={CellValue} />);
      const expected = '50 available at $10.00 each';

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('col');
      expect(rendered).toHaveText(expected);
    });
  });

  describe('with value: a component and cellProps: an object', () => {
    const cellProps = { promo: 'Special!' };
    const CellValue = (props) => {
      const { promo, price, quantity } = props;

      return (
        <span>{ promo } { quantity } available at ${ price } each</span>
      );
    };

    it('should render the row cell', () => {
      const rendered = shallow(
        <DynamicTableRowCell
          {...defaultProps}
          prop="price"
          value={CellValue}
          cellProps={cellProps}
        />,
      );
      const expected = 'Special! 50 available at $10.00 each';

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('col');
      expect(rendered).toHaveText(expected);
    });
  });

  describe('with value: a function', () => {
    const value = obj => `$${obj.price}`;

    it('should render the row cell', () => {
      const rendered = shallow(<DynamicTableRowCell {...defaultProps} prop="price" value={value} />);
      const expected = '$10.00';

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('col');
      expect(rendered).toHaveText(expected);
    });
  });

  describe('with value: a function and cellProps: an object', () => {
    const cellProps = { promo: 'Sale!' };
    const value = obj => `${obj.promo} $${obj.price}`;

    it('should render the row cell', () => {
      const rendered = shallow(
        <DynamicTableRowCell
          {...defaultProps}
          prop="price"
          value={value}
          cellProps={cellProps}
        />,
      );
      const expected = 'Sale! $10.00';

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('col');
      expect(rendered).toHaveText(expected);
    });
  });

  describe('with width: value', () => {
    it('should render the row cell', () => {
      const rendered = shallow(<DynamicTableRowCell {...defaultProps} width={3} />);
      const { name } = data;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('dynamic-table-row-cell');
      expect(rendered).toHaveClassName('col-3');
      expect(rendered).toHaveText(name);
    });
  });
});
