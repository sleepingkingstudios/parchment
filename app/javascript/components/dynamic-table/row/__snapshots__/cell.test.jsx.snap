// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicTableRowCell /> with default properties should match the snapshot 1`] = `
<div
  className="dynamic-table-row-cell col"
>
  Self-sealing Stem Bolt
</div>
`;
